editors/code/src/inlay_hints.ts

```diff
@@ -1,162 +1,319 @@
         } else {
-            return this.clear();
+            this.clearHints();
         }
     }
 
-    clear() {
-        this.allEditors.forEach(it => {
-            this.setTypeDecorations(it, []);
-            this.setParameterDecorations(it, []);
-        });
+    clearHints() {
+        for (const file of this.sourceFiles) {
+            file.inlaysRequest?.cancel();
+            file.renderHints([], this.client.protocol2CodeConverter);
+        }
     }
 
-    async refresh() {
+    refreshRustDocument(document: RustTextDocument) {
         if (!this.enabled) return;
-        await Promise.all(this.allEditors.map(it => this.refreshEditor(it)));
+
+        const file = this.sourceFiles.getSourceFile(document.uri.toString());
+
+        assert(!!file, "Document must be opened in some text editor!");
+
+        void file.fetchAndRenderHints(this.client);
     }
 
-    private get allEditors(): vscode.TextEditor[] {
-        return vscode.window.visibleTextEditors.filter(
-            editor => editor.document.languageId === 'rust',
-        );
+    refreshVisibleRustEditors(visibleEditors: RustTextEditor[]) {
+        if (!this.enabled) return;
+
+        const visibleSourceFiles = this.sourceFiles.drainEditors(visibleEditors);
+
+        // Cancel requests for source files whose editors were disposed (leftovers after drain).
+        for (const { inlaysRequest } of this.sourceFiles) inlaysRequest?.cancel();
+
+        this.sourceFiles = visibleSourceFiles;
+
+        for (const file of this.sourceFiles) {
+            if (!file.rerenderHints()) {
+                void file.fetchAndRenderHints(this.client);
+            }
+        }
     }
+}
+
 
-    private async refreshEditor(editor: vscode.TextEditor): Promise<void> {
-        const newHints = await this.queryHints(editor.document.uri.toString());
-        if (newHints == null) return;
-
-        const newTypeDecorations = newHints
-            .filter(hint => hint.kind === ra.InlayKind.TypeHint)
-            .map(hint => ({
-                range: this.ctx.client.protocol2CodeConverter.asRange(hint.range),
-                renderOptions: {
-                    after: {
-                        contentText: `: ${hint.label}`,
-                    },
-                },
-            }));
-        this.setTypeDecorations(editor, newTypeDecorations);
-
-        const newParameterDecorations = newHints
-            .filter(hint => hint.kind === ra.InlayKind.ParameterHint)
-            .map(hint => ({
-                range: this.ctx.client.protocol2CodeConverter.asRange(hint.range),
-                renderOptions: {
-                    before: {
-                        contentText: `${hint.label}: `,
-                    },
-                },
-            }));
-        this.setParameterDecorations(editor, newParameterDecorations);
+/**
+ * This class encapsulates a map of file uris to respective inlay hints
+ * request cancellation token source (cts) and an array of editors.
```

Comment:
Haven't read into this yet, but I think it should be possible to solve this without storing editors. I would think that the current `queryHints` logic is enough here, it should handle per-document in-flight requests correctly (modulo bugs). 

It seems it should be enought to tweak `refresh` function logic such that it first groups editors by the document they are displaying, and then refreshes each group, without maintaing any additional state or datastructures.

Reply:
The state keeps track of editors with already rendered inlay hints (some cache to prevent requests when opening duplicate editors) and prevents interaction with disposed editors (see 
[this line](https://github.com/rust-analyzer/rust-analyzer/pull/3378/files#diff-9994002c011d3bb124598d583d9fd8b1R124)) (vscode issues a warning for this otherwise).

Reply:
And it also prevent request for inlay hints when the user browses unrelated files (i.e. #3319)

Reply:
Would the following work as a way to get rid of the extra state here?

```TypeScript
class RustSourceFile {
    constructor(
        public readonly document: vscode.TextDocument,
        public inlaysRequest: null | vscode.CancellationTokenSource = null,
        public decorations: null | {
            type: vscode.DecorationOptions[];
            param: vscode.DecorationOptions[];
        } = null
    ) { }

    get editors(): vscode.TextEditor[] {
        return vscode.window.visibleTextEditors.filter((it) => it.document.uri.toString == this.document.uri.toString)
    }
}
```

Reply:
This *will* remove the extra state, though we will looze the info about which editors have decorations applied and which don't. So removing the state like this we will still rerender the decorations while the user is browsing editors/outputs/terminal widnows, but we will at least have `decorations` cache in order not to issue requests for inlay hints to the lsp server.

Reply:
I can think of dirty hack where we add boolean properties to `editors` to put the state in them instead of our custom map.

Reply:
> I can think of dirty hack where we add boolean property `decorationsApplied` to `editors` to put the state in them instead of our custom map.

Reply:
Or better use a [`symbol`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol). This will **do** work and maybe would be more efficient in terms of performance. However, the stability of a reference to the editor (i.e. pointer to it) is not documented by vscode, though in practice we can rely on this.

Reply:
I think the following example will make this more clear: 
```
const DEC = Symbol("decorations");

get editorToRender() {
    return vscode.window.visibleTextEditrs.fitler(suspect => suspect[DEC] && <uris equal>...)
}
```